### ISSUE TITLE
v2 Remix

### ISSUE THREAD

**Issue by cayuu** [mekanika/skematic#26]
Major goals of this remix:
1. **Simpler API**: v0.16 is super 👾 complex. Take things _out_
2. **Lower model impedance**: Closer to Postgres (see [sequelize](http://sequelizejs.com/))

**Comment by cayuu** [comment]
## API Design

> ⚠️ 🚧 **Work in Progress**

Would be handy to be able to preload Skematic with a model and opts to chain to subsequent data, eg.

``` js
const Player = Skematic(userModel, options)
const rambo = Player.format(data)
const dutch = Player.format(data, optOverrides)
const {error} = Skematic(gameModel, myOpts).validate(payload)
```

Or using statics, but with the optional parameter `opts` passed as the last argument:

#### `.format(model, data [, opts])`
- opts.

#### `validate(model, data [, opts])`

And notably, NO other API methods. No loading schemas or generators for reference by string. Code or bust.

**Comment by cayuu** [comment]
## Model Design

> ⚠️ 🚧 **Work in Progress**

As close as possible to Postgres and `sequelize`
- **type**: Needs to map cleanly to things like `VARCHAR(32)`.. 🚧 
- **allowNull**: For validating, accept `<type>` AND accept `null` ☑️ 
- ~~**default**: Consider enabling `fn()` at this level. eg. `Skm.NOW`(Note: decided to leave fn generators in the `generate` directive)~~

Enable mechanism to declare **indices** and handling on fields. From sequelize: 
- `unique: boolean|string` (where equal strings create composite keys)
- `primaryKey` ❗ Differs from v0.16 implementation (this one sets `NOT NULL UNIQUE`)
- `autoIncrement`

Other interesting maps:
- `field`: Specify a custom table name (map var to table)
- **references**: Create foreign key references (similar to v0.16 sub-schema)
  - `key`: The key to id on
  - `model`: Schema
  - `deferrable`: Postgres specific lookup timing cue

**Comment by cayuu** [comment]
Removed `$dynamic` keys support (attaching schema models to arbitrary keys). ba2d03ece

**Comment by cayuu** [comment]
On the plus side, validation performance of v2 is **4x** faster than `0.16` 😁

**Comment by cayuu** [comment]
What remains:
- [x] Is `null` handling hitting all the necessary use cases? #12 (yes)
- [x] Should "halt on first error" be included in v2 #10 (not 2.0.0)

Shifting these into own tickets. Not 2.0.0 necessities:
- (Dealing with as 2+) What should the Type system really be doing #27 (validating)
- How/what **index** mechanisms #31
- Foreign keys/references #32

**Comment by cayuu** [comment]
Ooooookay. #27 #31 and #32 have been totally taken off the table, so hey you know what, v2 is done and actually once extracted, this should be Skematic 3.0.

Closing this because 2.0 is forked 🍴🍳
